### PR TITLE
Devito3.0: more and more features

### PR DESCRIPTION
**Pull request by FabioLuporini**
This PR brings Devito3.0 steps closer to feature completeness.

Support for running several Acoustic operators is introduced. This mainly consists of properly dealing with the DSE-produced temporaries during the various DLE passes as well as for code generation (eg, inferring the right data types). However, to run the Acoustic operators, we'll need #187 .

Performance profiling has been extended, improved, and fixed where necessary (fixes were minor).

On top of all this, a number of improvements of various nature have been introduced.

### CODE REVIEW

File: devito/stencilkernel.py [hunk 1]
```diff
@@ -299,6 +474,32 @@ def argtypes(self):
 }
 
 """
+StencilKernel options
+"""
+options = {
+    'at_squeezer': 3,
+    'at_blocksize': [8, 16, 32]
+}
```
Comment (mloubout) [inline devito/stencilkernel.py]: What's the reason for this sizes?

Comment (FabioLuporini) [inline devito/stencilkernel.py]: nothing particularly solid: they simply looked reasonable sizes given the expected size of the parallel loop and the expected number of threads. And in my initial experiments (in Devito3.0) they resulted in performance improvements, while slightly different block sizes didn't change much the situation. I'll introduce more options in the future.

File: examples/benchmark.py [hunk 2]
```diff
@@ -182,46 +192,56 @@ def run(self, *args, **kwargs):
             warning("Could not load any results, nothing to plot. Exiting...")
             sys.exit(0)
 
-        gflopss = bench.lookup(params=parameters, measure="gflopss", event="loop_body")
-        oi = bench.lookup(params=parameters, measure="oi", event="loop_body")
-        time = bench.lookup(params=parameters, measure="timings", event="loop_body")
+        gflopss = bench.lookup(params=parameters, measure="gflopss", event="main")
+        oi = bench.lookup(params=parameters, measure="oi", event="main")
+        time = bench.lookup(params=parameters, measure="timings", event="main")
```
Comment (mlange05) [inline examples/benchmark.py]: Is this backwards compatible? I'm assuming we want to run 2.0.1 vs. 3.0 in like-for-like performance comparisons, so should we have a switch here?

File: examples/benchmark.py [hunk 3]
```diff
@@ -170,7 +180,7 @@ def run(self, *args, **kwargs):
         bench = Benchmark(
             name=args.problem, resultsdir=args.resultsdir, parameters=parameters
         )
-        bench.execute(BenchExecutor(), warmups=0)
+        bench.execute(BenchExecutor(), warmups=0, repeats=1)
```
Comment (mlange05) [inline examples/benchmark.py]: Did you mean to commit that?

**Comment by mlange05** [comment]
Apart from the backward-compatibility in `benchmark.py` this looks good to me.

File: examples/benchmark.py (hunk 2 above)
Comment (FabioLuporini) [inline examples/benchmark.py]: It's not backward compatible for plotting, you're right. But I guess for comparison we're just interested in raw numbers, or at least something simpler to compare than roofline plots? Anyway, good point: I will add a switch

File: examples/benchmark.py (hunk 3 above)
Comment (FabioLuporini) [inline examples/benchmark.py]: nope, that was just for testing purposes. Good spot

File: examples/benchmark.py (hunk 2 above)
Comment (FabioLuporini) [inline examples/benchmark.py]: done

File: examples/benchmark.py (hunk 3 above)
Comment (FabioLuporini) [inline examples/benchmark.py]: done

**Comment by FabioLuporini** [comment]
@navjotk : I added the TODO as discussed yesterday

**Comment by mlange05** [comment]
All looks good. Merging...

**Pull request merged**